src/posix.c

```diff
@@ -182,15 +198,45 @@ int p_fallocate(int fd, off_t offset, off_t len)
 
 	/*
 	 * fcntl will often error when the file already exists; ignore
-	 * this error since ftruncate will also resize the file (although
+	 * this error since our fallback will also resize the file (although
 	 * likely slower).
 	 */
-	fcntl(fd, F_PREALLOCATE, &prealloc);
-
-	return ftruncate(fd, (offset + len));
+	if ((error = fcntl(fd, F_PREALLOCATE, &prealloc)) < 0)
+		goto fallback;
+#elif defined(GIT_USE_FALLOCATE)
+	do {
+		error = fallocate(fd, 0, offset, len);
+	} while (error < 0 && errno == EINTR);
+	if (error < 0 && errno == ENOTSUP)
+		goto fallback;
 #else
-	return posix_fallocate(fd, offset, len);
+	goto fallback;
 #endif
+
+fallback:
```

Comment:
I don't quite get it. Aren't we now _always_ executing the fallback, regardless of whether the platform-specific implementation succeeded or not?

Reply:
Heh, true, I seem to have forgot a few `returns` when whatever platform-specific version succeeded. IOW, only failures of the (potential) platform fallocate would cause the fallback to run.